src/LiveComponent/assets/src/Component/index.ts

```diff
@@ -122,11 +126,31 @@ export default class Component {
      *     * loading.state:finished (element: HTMLElement) => {}
      *     * model:set (model: string, value: any, component: Component) => {}
      */
-    on(hookName: string, callback: (...args: any[]) => void): void {
+    on(hookName: 'connect', callback: (component: Component) => MaybePromise): void;
```

Comment:
What do you think about adding callback types too? Like `ConnectHook` or `ConnectCallback`

Reply:
Just a suggestion, I'm not so comfortable with TypeScript, but what would you think about something like that?

```ts
type Hooks = {
    'connect': (component: Component) => MaybePromise,
    'disconnect': (component: Component) => MaybePromise,
    'request:started': (requestConfig: any) => MaybePromise
    'render:finished': (component: Component) => MaybePromise
    'response:error': (backendResponse: BackendResponse, controls: {displayError: boolean}) => MaybePromise
    'loading.state.started': (element: HTMLElement, request: BackendRequest) => MaybePromise
    'loading.state.finished': (element: HTMLElement) => MaybePromise
    'model:set': (model: string, value: any, component: Component) => MaybePromise,
}

type ComponentHookName = keyof Hooks;

type HookCallback<T extends string> = T extends ComponentHookName
    ? Hooks[T]
    : (...args: any[]) => MaybePromise;

class Component {
    on<T extends ComponentHookName|string>(name: T, callback: HookCallback<T>): void {
        this.hooks.register(hookName, callback);
    }
    off<T extends ComponentHookName|string>(name: T, callback: HookCallback<T>): void {
        this.hooks.unregister(hookName, callback);
    }
}


```

Reply:
Hi @squrious ! I tested this approach, and after a few tweaks, it gives the same result as the current implementation, and is a bit easier to use for maintaining many events, rather than overloading the methods manually. I personally wouldn't be opposed to it. :smile:

Reply:
Nice :smile:

Reply:
So I went ahead and implemented your idea with the conditional, it's way more maintainable and achieves the same thing. :+1: